mllam_data_prep/derived_variables.py

```diff
@@ -0,0 +1,459 @@
+import importlib
+import sys
+
+import numpy as np
+import xarray as xr
+from loguru import logger
+
+
+def derive_variables(fp, derived_variables, chunking):
+    """
+    Load the dataset, and derive the specified variables
+
+    Parameters
+    ---------
+    fp : str
+        Filepath to the source dataset, for example the path to a zarr dataset
+        or a netCDF file (anything that is supported by `xarray.open_dataset` will work)
+    derived_variables : dict
+        Dictionary with the variables to derive
+        with keys as the variable names and values with entries for
+        kwargs and function to use in the calculation
+    chunking: dict
+        Dictionary with keys as the dimensions to chunk along and values
+        with the chunk size
+
+    Returns
+    -------
+    ds : xr.Dataset
+        Dataset with derived variables included
+    """
+    logger.info("Deriving variables")
+
+    try:
+        ds = xr.open_zarr(fp)
+    except ValueError:
+        ds = xr.open_dataset(fp)
+
+    ds_subset = xr.Dataset()
+    ds_subset.attrs.update(ds.attrs)
+    for _, derived_variable in derived_variables.items():
+        required_kwargs = derived_variable.kwargs
+        function_name = derived_variable.function
+        derived_variable_attributes = derived_variable.attributes or {}
+
+        # Separate the lat,lon from the required variables as these will be derived separately
+        latlon_coords_to_include = {}
+        for k, v in list(required_kwargs.items()):
+            if k in ["lat", "lon"]:
+                latlon_coords_to_include[k] = required_kwargs.pop(k)
```

Comment:
This is so that we can remove the `lat` and `lon` variables since they should be retrieved based on the projection information and not directly taken from the input dataset in the future (#33). However, this is not a very general formulation (since I write `if k in ["lat", "lon"]`) since this requires that they are named `lat` and `lon` in the source dataset and not something else (like `latitude` and `longitude`). But I couldn't come up with something more general. Any ideas?

Reply:
I think this is fine for now :) We can wait and not merge your work here until #33 is complete, or we can merge your's first and then adapt with #33. I know @observingClouds is on holiday now, so maybe for now simply issue a warning to say that you are assuming that the lat/lon coordinates are given as variables called `lat` and `lon`

Reply:
Sounds good to me. I have added a warning for this.